Introduce modem trace backend interface (#6772)
nrfconnect/sdk-nrf: doc/nrf/libraries/modem/nrf_modem_lib.rst

For better load distribution on the application, enable the experimental thread-based trace processing through the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PROCESSING` Kconfig option.
Trace processing is done in ``trace_handler_thread``.
When the modem trace module receives trace data, it places it in a FIFO queue.
The thread reads from the FIFO queue and forwards the trace data to the configured trace backend.
The FIFO queue used by the modem trace module uses dedicated heap memory.
If the modem trace backend is unable to keep up with the modem traces, the heap size can be increased by enabling :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE_OVERRIDE` and configuring :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE`.

> Increasing the heap size allows more traces in the FIFO queue, but the trace heap will still be depleted if the modem continues to send traces at a rate faster than the rate at which the backend can handle over time.

Reviewer: ```suggestion
Increasing the heap size allows more traces in the FIFO queue. However, if the modem continues to send traces at a faster rate than the backend can handle, the trace heap will be exhausted over time.
```

Reviewer 2: Adding a minor suggestion on top of @eivindj-nordic  suggestion.
```suggestion
Increasing the heap size allows more traces in the FIFO queue.
However, if the modem continues to send traces at a faster rate than the backend can handle, the trace heap will be exhausted over time.
```

> If increasing the trace heap size does not help, either optimize the backend speed or use a faster trace backend.

Reviewer: ```suggestion
If increasing the trace heap size does not help, either optimize the backend throughput or use a faster trace backend.
```

.. _adding_custom_modem_trace_backends:

Adding custom trace backends
============================

> You can add a custom trace backends if the existing trace backends are not sufficient.

Reviewer: ```suggestion
You can add custom trace backends if the existing trace backends are not sufficient.
```

At any time, only one trace backend can be compiled with the application.

> The value of the kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND` Kconfig option determines which trace backend is compiled.

Reviewer 2: ```suggestion
The value of the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND` Kconfig option determines which trace backend is compiled.
```

The :ref:`modem_trace_backend` sample demonstrates how a custom trace backend can be added to an application.

Complete the following steps to add a custom trace backend:

1. Place the files that have the custom trace backend implementation in a library or an application you create.
For example, the implementation of the UART trace backend (default) can be found in the :file:`nrf/lib/nrf_modem_lib/trace_backends/uart/uart.c` file.

> 2. Add a C file implementing the interface declared in the :file:`nrf/include/modem/trace_backend.h` header file.
> 
>    .. code-block:: c
> 
>    /* my_trace_backend.c */
> 
>    #include <modem/trace_backend.h>
> 
>    int trace_backend_init(void)
>    {
>            /* initialize trace backend here */
>            return 0;
>    }
> 
>    int trace_backend_deinit(void)
>    {
>            /* optional deinitialization code here */
>            return 0;
>    }
> 
>    int trace_backend_write(const void *data, size_t len)
>    {
>            /* forward data over custom backend here */
>            /* return number of bytes written or negative error code on failure */
>            return 0;
>    }
> 

Reviewer 2: ```suggestion
#. Add a C file implementing the interface in :file:`nrf/include/modem/trace_medium.h` header file.
   
   .. code-block:: c
      
      /* my_trace_medium.c */
      #include <modem/trace_medium.h>
      int trace_medium_init(void)
      {
           /* initialize transport medium here */
           return 0;
      }
      int trace_medium_deinit(void)
      {
           /* optional deinitialization code here */
           return 0;
      }
      int trace_medium_write(const void *data, size_t len)
      {
           /* forward data over custom transport here */
           /* return number of bytes written or negative error code on failure */
      }
```
This should fix the doc build error. The code written should be in line with 'c' of the 'code-block'.

Increasing the heap size allows more traces in the FIFO queue, but the trace heap will still be depleted if the modem continues to send traces at a rate faster than the rate at which the medium can handle over time.
If increasing the trace heap size does not help, either optimize the medium speed or use a faster trace transport medium.

.. _adding_custom_modem_trace_transport_mediums:

Adding custom trace transport mediums
=====================================

If the existing trace transport mediums are not sufficient, custom transport mediums can be added.
Note that only one transport medium can be compiled with the application at any time.
Which trace medium that is compiled is controlled by the Kconfig choice :kconfig:option:`NRF_MODEM_LIB_TRACE_MEDIUM`.

Adding a custom transport medium can be done by following the steps below.
Place the custom transport medium implementation in a library or an application you create.
As a reference, the implementation of the UART trace medium (default medium) can be found in :file:`nrf/lib/nrf_modem_lib/trace_mediums`.
The :ref:`custom_trace_medium_sample` sample demonstrates how a custom trace medium can be added to an application.

> 1. Add a C file implementing the interface in :file:`nrf/include/modem/trace_medium.h`\.
>    .. code-block:: c
>    /* my_trace_medium.c */
>    #include <modem/trace_medium.h>
> 
>    int trace_medium_init(void)
>    {
>            /* initialize transport medium here */
>            return 0;
>    }
> 
>    int trace_medium_deinit(void)
>    {
>            /* optional deinitialization code here */
>            return 0;
>    }
> 
>    int trace_medium_write(const void *data, size_t len)
>    {
>            /* forward data over custom transport here */
>            /* return number of bytes written or negative error code on failure */
>    }

Reviewer 2: ```suggestion
#. Add a C file implementing the interface in :file:`nrf/include/modem/trace_medium.h` header file.
   
   .. code-block:: c
      
      /* my_trace_medium.c */
      #include <modem/trace_medium.h>
      int trace_medium_init(void)
      {
           /* initialize transport medium here */
           return 0;
      }
      int trace_medium_deinit(void)
      {
           /* optional deinitialization code here */
           return 0;
      }
      int trace_medium_write(const void *data, size_t len)
      {
           /* forward data over custom transport here */
           /* return number of bytes written or negative error code on failure */
      }
```

> #. Add or modify a :file:`Kconfig` file, extending the choice :kconfig:option:`NRF_MODEM_LIB_TRACE_MEDIUM` with another option.
>    .. code-block:: Kconfig
>       if NRF_MODEM_LIB_TRACE_ENABLED
> 
>       # Extends the choice with another medium
>       choice NRF_MODEM_LIB_TRACE_MEDIUM
> 
>       config NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM
>               bool "My trace transport medium"
>               help
>                 Optional description of my
>                 trace transport medium.
> 
>       endchoice
> 
>       endif

Reviewer 2: ```suggestion
#. Create or modify a :file:`Kconfig` file to set the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM` Kconfig option with another option.

   .. code-block:: Kconfig
   
      if NRF_MODEM_LIB_TRACE_ENABLED
      
      # Extends the choice with another medium
      choice NRF_MODEM_LIB_TRACE_MEDIUM  

      config NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM
             bool "My trace transport medium"
             help
              Optional description of my
              trace transport medium.

      endchoice

      endif
```

Reviewer 3: @divipillai This particular Kconfig code is not setting the option but extending it

> #. Add or modify a :file:`CMakeLists.txt` file, adding the custom trace medium source(s) only if the custom trace medium option have been chosen.
>    .. code-block:: cmake
>       if(CONFIG_NRF_MODEM_LIB_TRACE_ENABLED)
> 
>       zephyr_library()
> 
>       # Only add 'custom' medium to compilation when selected.
>       zephyr_library_sources_ifdef(
>         CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM
>         path/to/my_trace_medium.c
>       )
> 
>       endif()

Reviewer 2: ```suggestion
#. Create or modify a :file:`CMakeLists.txt` file, adding one or more custom trace medium sources only if the custom trace medium option have been chosen.

   .. code-block:: cmake
   
      if(CONFIG_NRF_MODEM_LIB_TRACE_ENABLED)

      zephyr_library()

      # Only add 'custom' medium to compilation when selected.
      zephyr_library_sources_ifdef(
        CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM
        path/to/my_trace_medium.c
      )

      endif()
```

> #. Make sure the :file:`Kconfig` and :file:`CMakeLists.txt` files are included in the build.
> #. To use the custom modem trace transport medium, add
>    .. code-block:: none
>       CONFIG_NRF_MODEM_LIB_TRACE_ENABLED=y
>       CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM=y
> 

Reviewer 2: ```suggestion
#. Include the :file:`Kconfig` file and the :file:`CMakeLists.txt` file to the build.
#. Add the following Kconfig options to your application's :file:`prj.conf` file to use the custom modem trace transport medium:
   
   .. code-block:: none
      
      CONFIG_NRF_MODEM_LIB_TRACE_ENABLED=y
      CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_MY_TRACE_MEDIUM=y
```
